pre_commit/languages/download.py

```diff
@@ -0,0 +1,318 @@
+
+        hasher = hashlib.new(self.algorithm)
+
+        checksum_len = len(b64decode(self.checksum))
+        if checksum_len != hasher.digest_size:
+            raise ValueError(
+                f'Invalid checksum string length of {checksum_len} for \
+{self.algorithm}, expected {hasher.digest_size}',
+            )
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @property
+    def algorithm(self) -> str:
+        return self.__algorithm
+
+    @property
+    def checksum(self) -> str:
+        return self.__checksum
+
+    def __str__(self) -> str:
+        return self.value
+
+    def check(self, io: HasBinaryRead, chunk: int = 4096) -> Iterator[bytes]:
+        hasher = hashlib.new(self.algorithm)
+        while buffer := io.read(chunk):
+            hasher.update(buffer)
+            yield buffer
+        digest = b64encode(hasher.digest()).decode('utf-8)')
+        if digest != self.checksum:
+            raise ChecksumMismatchError(
+                expected=self,
+                actual=SRI(f'{self.algorithm}-{digest}'),
+            )
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, SRI):
+            return False
+        return (
+            other.algorithm == self.algorithm and
+            other.checksum == self.checksum
+        )
+
+
+@dataclass
+class URI:
+    value: str
+
+    def __post_init__(self) -> None:
+        url = urlparse(self.value)
+        if not all([url.scheme, url.netloc]):
+            raise ValueError(f'Invalid URI: {self.value}')
+        self._netloc = url.netloc
+
+    @property
+    def netloc(self) -> str:
+        return self._netloc
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class Metadata:
+    value: str
+
+    @property
+    def parts(self) -> tuple[str, str, str, str]:
+        first, second, third, fourth = self.value.splitlines()
+        return (first, second, third, fourth)
+
+    @property
+    def platform(self) -> Platform:
+        platform, _, _, _ = self.parts
+        return Platform(platform)
+
+    @property
+    def sri(self) -> SRI:
+        _, sri, _, _ = self.parts
+        return SRI(sri)
+
+    @property
+    def uri(self) -> URI:
+        _, _, uri, _ = self.parts
+        return URI(uri)
+
+    @property
+    def filename(self) -> PurePath:
+        _, _, _, path = self.parts
+        return PurePath(path)
+
+
+def download(uri: URI, sri: SRI, filename: Path) -> None:
+    request = Request(str(uri))
+    manager = HTTPPasswordMgrWithDefaultRealm()
+    handler = HTTPBasicAuthHandler(manager)
+    opener = build_opener(handler)
+    try:
+        authenticators = netrc().authenticators(uri.netloc)
+        if authenticators is not None:
+            login, _, password = authenticators
+            if password is None:
+                password = ''
+            manager.add_password(None, str(uri), login, password)
+    except FileNotFoundError:
```

Comment:
``netrc`` file doesn't have to exist. It is reasonable for it not to exist. None of the other apps I have used reports an error message when ``netrc`` file is missing.

Reply:
Agreed. You may want to scope the exception net tighter to indicate what code you want to catch the exception for:

```py
try:
    nc = netrc()
except FileNotFoundError:
    pass
else:
    auth = nc.authenticators(uri.netloc)
    # ...add the auth...
```

That _clearly_ implies that you _only_ want to catch the exception for the `netrc` call and nothing else.

Reply:
Done